src/functional/iterator/transforms/unroll_reduce.py

```diff
@@ -1,11 +1,98 @@
 import dataclasses
-from collections.abc import Iterable
+from collections.abc import Iterable, Iterator
+from typing import TypeGuard
 
 from eve import NodeTranslator
 from eve.utils import UIDGenerator
+from functional import common
 from functional.iterator import ir
 
 
+def _is_shifted(arg: ir.Expr) -> TypeGuard[ir.FunCall]:
+    return (
+        isinstance(arg, ir.FunCall)
+        and isinstance(arg.fun, ir.FunCall)
+        and arg.fun.fun == ir.SymRef(id="shift")
+    )
+
+
+def _is_applied_lift(arg: ir.Expr) -> TypeGuard[ir.FunCall]:
```

Comment:
I like the name. We should rename / unify with the equivalents in the lift inliner.